# Laboratory Session 1
## Python Practice Exercises

In [1]:
import pandas as pd
import numpy as np
from itertools import combinations as C

### Exercise 1:
Write a program which will find all such numbers which are symetric (such as 11, 121, 1221,...) between 10 and 2000 (both included). The numbers obtained should be printed in a commaseparated sequence on a single line.

In [2]:
def symnum(a, b):
    if (type(a)!=int) or (type(b)!=int):
        print('Please enter two integers.')
    else:
        i=a
        while(i<(b+1)):
            m=i
            n=i
            rev=0
            while(n>0):
                ld=n%10
                rev=rev*10+ld
                n=n//10
            if(m==rev):
                print(i, end=', ')
            i=i+1

In [3]:
symnum(10, 2000)

11, 22, 33, 44, 55, 66, 77, 88, 99, 101, 111, 121, 131, 141, 151, 161, 171, 181, 191, 202, 212, 222, 232, 242, 252, 262, 272, 282, 292, 303, 313, 323, 333, 343, 353, 363, 373, 383, 393, 404, 414, 424, 434, 444, 454, 464, 474, 484, 494, 505, 515, 525, 535, 545, 555, 565, 575, 585, 595, 606, 616, 626, 636, 646, 656, 666, 676, 686, 696, 707, 717, 727, 737, 747, 757, 767, 777, 787, 797, 808, 818, 828, 838, 848, 858, 868, 878, 888, 898, 909, 919, 929, 939, 949, 959, 969, 979, 989, 999, 1001, 1111, 1221, 1331, 1441, 1551, 1661, 1771, 1881, 1991, 

### Exercise 2:
Write a program which will find all such numbers which include at least 3 duplicated digits (such as 111, 1222,...) between 100 and 2000 (both included). The numbers obtained should be printed in a comma-separated sequence on a single line.

In [4]:
def duplicated_digit(num):
    num_list=list(str(num))
    unique_value= np.unique(np.array(num_list))
    count_num =[]
    for i in unique_value:
        count_num.append(num_list.count(i))
    max_num=max(count_num)
    return_dict= {'distinct value':unique_value,'count':count_num, 'max': max_num}
    return return_dict

In [5]:
for i in np.linspace(100, 2000, 1901):
    if duplicated_digit(int(i))['max']>=3:
        print(int(i),end=', ')

111, 222, 333, 444, 555, 666, 777, 888, 999, 1000, 1011, 1101, 1110, 1111, 1112, 1113, 1114, 1115, 1116, 1117, 1118, 1119, 1121, 1131, 1141, 1151, 1161, 1171, 1181, 1191, 1211, 1222, 1311, 1333, 1411, 1444, 1511, 1555, 1611, 1666, 1711, 1777, 1811, 1888, 1911, 1999, 2000, 

### Exercise 3:
Write a program which will receive an integer larger than 0 and respond number of digits of that integer.

In [6]:
def lenint(a):
    if (type(a)!=int) or (a<=0):
        print('Please enter an integer larger than 0.')
    else:
        print('Number of digits: ', len(list(str(a))))

In [7]:
lenint(2508)

Number of digits:  4


### Exercise 4:
Write a program which will receive an integer larger than 0 and present that integer in binary format.

In [8]:
def binform(a):
    if (type(a)!=int) or (a<=0):
        print('Please enter an integer larger than 0.')
    else:
        print('Binary format: ', format(a, 'b'))

In [9]:
binform(2508)

Binary format:  100111001100


### Exercise 5:
Write a program which will receive an integer larger than 100 and find all divisors of that integer. The numbers obtained should be printed in a comma-separated sequence on a single line.

In [10]:
def divisors(a):
    if (type(a)!=int) or (a<=100):
        print('Please enter an iteger larger than 100.')
    else:
        print('The divisors of ', a,': ' )
        for i in range(1, a+1):
            if a%i==0:
                print(i, end=', ')

In [11]:
divisors(2508)

The divisors of  2508 : 
1, 2, 3, 4, 6, 11, 12, 19, 22, 33, 38, 44, 57, 66, 76, 114, 132, 209, 228, 418, 627, 836, 1254, 2508, 

### Exercise 6:
Write a program which will find all primes between 1000 and 3000 (both included). The numbers obtained should be printed in a comma-separated sequence on a single line. 

In [15]:
def primes(a, b):
    if (type(a)!=int) or (type(b)!=int) or (b<=a):
        print('Please enter two integers in ascending order.')
    else:
        print('The prime between ', a, ' and ', b, ': ')
        for i in range(a, b+1):
            prime=True
            for j in range(2, i):
                if(i%j==0):
                    prime=False
            if prime:
                print(i, end=', ')

In [16]:
primes(1000, 3000)

The prime between  1000  and  3000 : 
1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181, 1187, 1193, 1201, 1213, 1217, 1223, 1229, 1231, 1237, 1249, 1259, 1277, 1279, 1283, 1289, 1291, 1297, 1301, 1303, 1307, 1319, 1321, 1327, 1361, 1367, 1373, 1381, 1399, 1409, 1423, 1427, 1429, 1433, 1439, 1447, 1451, 1453, 1459, 1471, 1481, 1483, 1487, 1489, 1493, 1499, 1511, 1523, 1531, 1543, 1549, 1553, 1559, 1567, 1571, 1579, 1583, 1597, 1601, 1607, 1609, 1613, 1619, 1621, 1627, 1637, 1657, 1663, 1667, 1669, 1693, 1697, 1699, 1709, 1721, 1723, 1733, 1741, 1747, 1753, 1759, 1777, 1783, 1787, 1789, 1801, 1811, 1823, 1831, 1847, 1861, 1867, 1871, 1873, 1877, 1879, 1889, 1901, 1907, 1913, 1931, 1933, 1949, 1951, 1973, 1979, 1987, 1993, 1997, 1999, 2003, 2011, 2017, 2027, 2029, 2039, 2053, 2063, 2069, 2081, 2083, 2087, 2089, 2099, 2111, 2113, 2129, 2131, 2137, 2141, 2143, 2153, 2161, 2179, 2203, 22

### Exercise 7:
Write a program which will receive an integer larger than 0 and respond all perfect numbers smaller than that integer. A perfect number is sum of its divisors except itself, such as 6 = 1+2+3. The numbers obtained should be printed in a comma-separated sequence on a single line. 

In [17]:
def perfnum(x):
    if(type(x)!=int) or (x<=0):
        print('Please enter an iteger larger than 0.')
    else:
        n=1
        while(n<x):
            sum1=0
            for i in range(1, n):
                if(n%i==0):
                    sum1=sum1+i
            if(sum1==n):
                print(n, end=', ')
            n=n+1

In [18]:
perfnum(2508)

6, 28, 496, 

### Exercise 8:
With a given integral number n, write a program to generate a dictionary that contains (i, i^3) such that is an integral number between 1 and n (both included). and then the program should print the dictionary. Suppose the following input is supplied to the program: 6. Then, the output should be: {1: 1, 2: 8, 3: 27, 4: 64, 5: 125, 6: 216}

In [23]:
def dupdict(x):
    if(type(x)!=int) or (x<=1):
        print('Please enter an iteger larger than 1.')
    else:
        dict1={}
        for i in range(1, x+1):
            dict1[i]=int(i)**3
        print(dict1)

In [24]:
dupdict(13)

{1: 1, 2: 8, 3: 27, 4: 64, 5: 125, 6: 216, 7: 343, 8: 512, 9: 729, 10: 1000, 11: 1331, 12: 1728, 13: 2197}


### Exercise 9:
Write a program which accepts a sequence of comma separated 4 digit binary numbers as its input and then check whether they are divisible by 3 or not. The numbers that are divisible by 3 are to be printed in a comma separated sequence. Example: 0100,0011,1010,1101 Then the output should be: 1101 

In [26]:
str1=str(input('Enter 4 comma-seperated binary numbers: '))
lst1=str1.split(',')
lst=[]
for i in lst1:
    lst.append(int(i))
print(lst)
j=0
print('The binary numbers divisible by 3: ')
while(j<len(lst)):
    if(lst[j]%3==0):
        print(lst[j], end=', ')
    j=j+1

Enter 4 comma-seperated binary numbers:  1001,1101,1011,1000

[1001, 1101, 1011, 1000]
The binary numbers divisible by 3: 
1101, 1011, 

### Exercise 10:
Write a program which will receive an integer larger than 0 and respond if that integer is sum of three distinct squared numbers, such as 38 = 22 + 32 + 52.

In [27]:
def sum_of_square(number):
    square=[]
    for i in range(1,(int(number**(1/2)))+1):
        square.append(i**2)
    combi_list=list(C(square,3))
    sum_n=[x+y+z for (x,y,z) in combi_list]
    n=0
    for i in range(len(sum_n)):
        if sum_n[i]==number:
            n=n+1
    return n

In [29]:
number=float(input('Enter a positive integer: '))
while number!= int(number) or number<= 0:
    number=float(input("Please enter a positive integer: "))
number=int(number)

if sum_of_square(number)==1:
    print("The input number is a sum of 3 distinct square numbers.")
else:
    print("The input number is not a sum of 3 distinct square numbers")

Enter a positive integer:  38

The input number is a sum of 3 distinct square numbers.
